app/policies/order/delivery_status_policy.rb

```diff
@@ -0,0 +1,23 @@
+class Order::DeliveryStatusPolicy
+
+  def initialize(user, order)
+    @user = user
+    @order = order
+  end
+
+  def mark_delivered?
+    !user.buyer_only? && order.undelivered_for_user?(user) &&
+      (user.can_manage_market?(order.market) || order.delivery.delivery_schedule.direct_to_customer?)
+  end
+
```

Comment:
I can can be easier to read if you extract some guards, something like:

    def mark_delivered?
      return false if user.buyer_only?
      return false unless order.undelivered_for_user?(user)
      user.can_manage_market? || delivery_is_direct_to_customer?
    end

    private    

    def delivery_is_direct_to_customer?
       order.delivery.delivery_schedule.direct_to_customer?
    end

Reply:
Cool, will do.